## 1.1. Sequência de Fibonacci

A Sequência de Fibonacci é uma sequência numérica de tamanho **N** onde qualquer número é dado pela soma dos dois elementos anteriores, com exeção do primeiro e do segundo número.


### 1.1.1. Primeira tentativa com recursão

In [ ]:
def fibonacci_1(n: int) -> int:
    return fibonacci_1(n - 1) + fibonacci_1(n - 2)

fibonacci_1(5)

: 

: 

O programa quebra pois não há uma condição de parada, então acabamos estourando a pilha de recursão do Python.

```python
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in fib
  File "<stdin>", line 2, in fib
  File "<stdin>", line 2, in fib
  [Previous line repeated 996 more times]
RecursionError: maximum recursion depth exceeded
```

Para solucionar isso, devemos trabalhar com uma condição de parada para a recursão.

### 1.1.2. Utilizando casos de base

O caso base de uma função recursiva é uma condição onde a função não chamará a si mesma novamente, quebrando assim a recursão e impedindo que aconteça um _loop infinito_.

In [5]:
fib_2_execution_count: int = 0

def fibonacci_2(n: int) -> int:
    global fib_2_execution_count
    fib_2_execution_count += 1
    # caso base que para a recursão
    if n < 2:
        return n
    # chamada recursiva
    return fibonacci_2(n - 2) + fibonacci_2(n - 1)

n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_2(n)}')
print(f'Para calculá-lo foram necessárias {fib_2_execution_count} chamadas a função fibonacci_2')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 2692537 chamadas a função fibonacci_2


### 1.1.3. Memoização

A memoização consiste em armazenar os resultados gerados por uma função, de forma que toda vez que formos executá-la novamente, caso os dados de entrada já tenham sido processados anteriormente, será retornado o valor armazenado, caso contrário a função será executada normalmente e ao final salvará a saída gerada para aquela entrada.

In [3]:
from typing import Dict

fib_3_execution_count: int = 0

# iniciamos nosso dicionário com os dois casos base da sequência de fibonacci
memo: Dict[int, int] = {0: 0, 1: 1}

def fibonacci_3(n: int) -> int:
    global fib_3_execution_count
    fib_3_execution_count += 1
    if n not in memo:
        memo[n] = fibonacci_3(n - 2) + fibonacci_3(n - 1)
    return memo[n]

n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_3(n)}')
print(f'Para calculá-lo foram necessárias {fib_3_execution_count} chamadas a função fibonacci_3')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 59 chamadas a função fibonacci_3


Ao utilizar a memoização, durante o cálculo do trigésimo número da sequência saimos de `2692537` chamadas a função `fibonacci_2` para apenas `59` chamadas a função `fibonacci_3`.

### 1.1.4. Memoização Automática

Para não precisar fazer a memoização manualmente, podemos usar o cache do [módulo functools do Python](https://docs.python.org/3/library/functools.html) através da anotação `@lru_cache`.

In [6]:
from functools import lru_cache

fib_4_execution_count: int = 0

@lru_cache
def fibonacci_4(n: int) -> int:
    global fib_4_execution_count
    fib_4_execution_count += 1

    if n < 2:
        return n
    # chamada recursiva
    return fibonacci_4(n - 1) + fibonacci_4(n - 2)

n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_4(n)}')
print(f'Para calculá-lo foram necessárias {fib_4_execution_count} chamadas a função fibonacci_4')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 31 chamadas a função fibonacci_4


Com isso, conseguimos diminuir o número de chamadas a nossa função de `59` em `fibonacci_3` para `31` em `fibonacci_4`.

### 1.1.5. Função Iterativa

Um outra abordagem para calcular os elementos da Sequência de Fibonacci é criando uma função iterativa, o que significa que ao invés de ter um grande número de chamadas recursivas ocupando nossa memória teremos apenas um laço de repetição que executará `n - 1` iterações.

In [9]:
fib_5_execution_count: int = 0

def fibonacci_5(n: int) -> int:
    if n == 0:
        return 0

    global fib_5_execution_count
    anterior: int = 0
    proximo: int = 1
    i: int = 1

    while(i < n):
        aux = anterior
        anterior = proximo
        proximo = anterior + aux

        fib_5_execution_count += 1
        i += 1

    return proximo


n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_5(n)}')
print(f'Para calculá-lo foram necessárias {fib_5_execution_count} iterações na função fibonacci_5')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 29 iterações na função fibonacci_5


Através da sintaxe simples e alguns recursos da linguagem Python, podemos reescrever essa função de uma maneira `"Pythonica"`:

In [10]:
fib_pythonico_execution_count: int = 0

def fibonacci_pythonico(n: int) -> int:
    if n == 0:
        return 0

    global fib_pythonico_execution_count
    anterior: int = 0
    proximo: int = 1

    for i in range(1, n):
        anterior, proximo = proximo, anterior + proximo
        fib_pythonico_execution_count += 1

    return proximo


n = 30
print(f'O {n} elemento da Sequência de Fibonacci é {fibonacci_pythonico(n)}')
print(f'Para calculá-lo foram necessárias {fib_pythonico_execution_count} iterações na função fibonacci_pythonico')

O 30 elemento da Sequência de Fibonacci é 832040
Para calculá-lo foram necessárias 29 iterações na função fibonacci_pythonico


### 1.1.6. Usando gerador (yield)

O Python possui o yield, um `generator`, que cria uma lista de dados que é consumida sob demanda, sendo utilizado para omitir a estrutura que gera os dados.

In [1]:
from typing import Generator

fib_6_execution_count: int = 0

def fibonacci_6(n: int) -> Generator[int, None, None]:
    yield 0
    if n > 0:
        yield 1

    global fib_6_execution_count
    anterior: int = 0
    proximo: int = 1
    i: int = 1

    while(i < n):
        aux = anterior
        anterior = proximo
        proximo = anterior + aux
        yield proximo

        fib_6_execution_count += 1
        i += 1


n = 30
for i in fibonacci_6(n):
    print(i, end=', ')
print('')

0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 
